File: 2_medium/3001_4000/03619_count_islands_with_total_value_divisible_by_k/dfs_recursive.cpp
class Solution {
    long dfs_recursive(vector<vector<int>> &grid, size_t i, size_t j, size_t rows, size_t cols) {

        if (i >= rows || j >= cols || grid[i][j] == 0) {
            return 0;
        }

        long sum = grid[i][j];
        grid[i][j] = 0;

        sum += dfs_recursive(grid, i - 1, j, rows, cols);
        sum += dfs_recursive(grid, i + 1, j, rows, cols);
        sum += dfs_recursive(grid, i, j - 1, rows, cols);
        sum += dfs_recursive(grid, i, j + 1, rows, cols);

        return sum;
    }

public:
    int countIslands(vector<vector<int>>& grid, int k) {
        int count = 0;
        size_t rows = grid.size();
        size_t cols = grid[0].size();

        for (size_t i{0}; i < rows; ++i) {
            for (size_t j{0}; j < cols; ++j) {
                if (grid[i][j] != 0) {
                    long sum = dfs_recursive(grid, i, j, rows, cols);

                    if (sum % k == 0) {
                        count += 1;
                    }
                }
            }
        }

        return count;
    }
};

File: 2_medium/3001_4000/03619_count_islands_with_total_value_divisible_by_k/dfs_recursive.rs
impl Solution {
    fn dfs_recursive(grid: &mut [Vec<i32>], i: usize, j: usize, rows: usize, cols: usize) -> i32 {

        if grid[i][j] == 0 {
            return 0;
        }

        let mut sum = grid[i][j];
        grid[i][j] = 0;

        if i > 0 {
            sum += Self::dfs_recursive(grid, i - 1, j, rows, cols);
        }
        if i < rows - 1 {
            sum += Self::dfs_recursive(grid, i + 1, j, rows, cols);
        }

        if j > 0 {
            sum += Self::dfs_recursive(grid, i, j - 1, rows, cols);
        }
        if j < cols - 1 {
            sum += Self::dfs_recursive(grid, i, j + 1, rows, cols);
        }

        sum
    }

    pub fn count_islands(mut grid: Vec<Vec<i32>>, k: i32) -> i32 {

        let mut count = 0;
        let rows = grid.len();
        let cols = grid[0].len();

        for i in 0..rows {
            for j in 0..cols {
                if grid[i][j] != 0 {

                    let mut sum = Self::dfs_recursive(&mut grid, i, j, rows, cols);

                    if sum % k == 0 {
                        count += 1;
                    }
                }
            }
        }

        count
    }
}
